packages/hdwallet-metamask-shapeshift-multichain/src/adapter.ts

```diff
@@ -72,39 +71,33 @@ const METAMASK_IMPERSONATORS = [
   "isVision",
 ];
 
+const store = createStore();
+
 export class MetaMaskAdapter {
   keyring: core.Keyring;
+  providerRdns: string;
 
-  private constructor(keyring: core.Keyring) {
+  private constructor(keyring: core.Keyring, providerRdns: string) {
     this.keyring = keyring;
+    this.providerRdns = providerRdns;
   }
 
-  public static useKeyring(keyring: core.Keyring) {
-    return new MetaMaskAdapter(keyring);
+  public static useKeyring(keyring: core.Keyring, providerRdns: string) {
+    return new MetaMaskAdapter(keyring, providerRdns);
   }
 
   public async initialize(): Promise<number> {
     return Object.keys(this.keyring.wallets).length;
   }
 
   public async pairDevice(): Promise<MetaMaskShapeShiftMultiChainHDWallet | undefined> {
-    // TODO: remove casting, @metamask/detect-provider npm package hasn't been published with latest typings
-    // https://github.com/MetaMask/detect-provider/blame/5ce916fc24779c4b36741531a41d9c8b3cbb0a17/src/index.ts#L37
-    const provider = (await detectEthereumProvider({
-      mustBeMetaMask: true,
-      silent: false,
-      timeout: 3000,
-    })) as providers.ExternalProvider | null;
-    if (!provider) {
-      const onboarding = new MetaMaskOnboarding();
-      onboarding.startOnboarding();
-      console.error("Please install MetaMask!");
-      throw new Error("MetaMask provider not found");
-    }
+    const maybeEip6963Provider = store.findProvider({ rdns: this.providerRdns });
```

Comment:
Yep, that's all it takes. Didn't even need to pass it from web in the end (which didn't work, something something proxies?), which is actually good, in-place provider detection with rdns

Reply:
Note naming: EIP-6963 structure here, with `provider` (an EIP-1193 compliant provider) and `info` (provider info) vs. 1193 below, the good ol' JS provider API https://eips.ethereum.org/EIPS/eip-1193